runtime/doc/lsp.txt

```diff
@@ -698,6 +715,75 @@ commands                                                    *vim.lsp.commands*
 
     The second argument is the `ctx` of |lsp-handler|
 
+config({name}, {cfg})                                       *vim.lsp.config()*
+    Define default configuration for LSP clients.
+
+    Use name '*' to set default configuration for all clients.
+
+    Can also be used as a table.
+
+    Examples:
+    • Set default root marker for all clients: >lua
+       vim.lsp.config('*', {
+         root_markers = { '.git' },
+       })
+<
+    • Set default configuration for clangd using table form: >lua
+       vim.lsp.config.clangd = {
+         cmd = {
+           'clangd',
+           '--clang-tidy',
+           '--background-index',
+           '--offset-encoding=utf-8',
+         },
+         markers = { '.clangd', 'compile_commands.json' },
+         filetypes = { 'c', 'cpp' },
+       }
+<
+    • Get configuration for luals: >lua
+       local cfg = vim.lsp.config.luals
+<
+
+    Parameters: ~
+      • {name}  (`string`)
+      • {cfg}   (`vim.lsp.Config`) See |vim.lsp.Config|.
+
+enable({name}, {config})                                    *vim.lsp.enable()*
+    Enable a LSP server for filetype(s) to automatically start when opening a
+    buffer.
+
+    Uses default configuration defined with `vim.lsp.config`.
+
+    Examples: >lua
+          vim.lsp.enable('clangd', {
+            cmd = { 'clangd', '--clang-tidy' },
+            root_markers = { '.clangd', 'compile_commands.json' },
```

Comment:
For reference, in https://github.com/neovim/nvim-lspconfig/issues/1827 it sounds like tpope was infavor of this "declarative root_markers". Is this derived from that?

Whereas https://github.com/neovim/neovim/pull/18506#issuecomment-1121958744 was hesitant: 

>  root_markers as a list of strings should not be a baked in concept like this, there are too many special cases that require more complicated logic (nested git projects, rust, etc). I don't think it's good design to have essentially duplicate overlap with root_dir (which should be removed IMO in favor of the more generic workspaceFolders adopted by the spec)

Reply:
I think `vim.fs.root` is the compromise we settled on? Allows specifying "declarative root markers" with minimal boilerplate while still allowing fully custom logic where needed.

Reply:
AFAIK this satisfies the large majority of use cases. If there are more complex cases then we can try to accommodate that in some way (with root_marker as function?), or defer them to use `vim.lsp.start()`.

At this point I think lspconfig has done enough baking; so we've got to be able to make progress, otherwise it will never happen.

To me saying root markers isn't a good enough general solution because of a few use cases is like saying treesitter isn't good enough because of a few languages.

Reply:
Yes, all these comments were before we had `vim.lsp.start` in core, which is the "full control" interface (and working fine for that). This is a (hopefully thin) additional layer for specifying "static configs" which does not need to have the full expressivity. 

(That being said, allowing `root_markers` to be a function doesn't hurt -- unless someone again starts raving about "fully declarative configuration" -- but in no way needs to block the first implementation here. Let's start with the "90% at 10%" cases.)

Reply:
Perfect. We have closed the loop here.